Added NoVNC console page

bond95 opened the pull request:
Fixes: #490 
Fixes: #966

 - Refactored Mike's code
 - Fixed console page
 - Improved NoVNC component
 - Changed `ConsoleConfirmationModal` architecture
 - Added `Loader` component
 - Added console info modal

Patterfly-react PR: https://github.com/patternfly/patternfly-react/pull/1402

![image](https://user-images.githubusercontent.com/3332176/52848545-4c8b9f80-310f-11e9-92a5-0aa02b9f2405.png)

![image](https://user-images.githubusercontent.com/3332176/52848569-5ca37f00-310f-11e9-9c71-8e236645ab01.png)

![image](https://user-images.githubusercontent.com/3332176/52848601-71801280-310f-11e9-9695-02487075f2b1.png)

![image](https://user-images.githubusercontent.com/3332176/52848634-7fce2e80-310f-11e9-9849-7e41e0ae7467.png)

![image](https://user-images.githubusercontent.com/3332176/52848660-8fe60e00-310f-11e9-9541-f7d3e474b740.png)

lwrigh commented:
What do you guys think about the possibility of adding a full screen button into the console page?  We could place it in the lower right hand corner of the console screen.

gregsheremeta commented:
> What do you guys think about the possibility of adding a full screen button into the console page? We could place it in the lower right hand corner of the console screen.

I'm not sure the novnc component supports full screen mode. @bond95 ?

bond95 commented:
@gregsheremeta well, yeah, component doesn't have any props for it, but it doesn't mean that we can't add it. But I think it would be better to do in separate PR.

gregsheremeta commented:
> @gregsheremeta well, yeah, component doesn't have any props for it, but it doesn't mean that we can't add it. But I think it would be better to do in separate PR.

Agree. I was just wondering if novnc supports it at all.

bond95 commented:
@lwrigh added icons 
![image](https://user-images.githubusercontent.com/3332176/53085049-60197a80-3502-11e9-93eb-65f0f35b8309.png)

lwrigh commented:
> @lwrigh added icons
> ![image](https://user-images.githubusercontent.com/3332176/53085049-60197a80-3502-11e9-93eb-65f0f35b8309.png)

Looks good to me 👍

sjd78 commented on src/components/Toolbar/index.js:
```diff
@@ -42,7 +47,40 @@ const PoolDetailToolbarConnected = connect(
+  if (vms.getIn(['vms', match.params.id])) {
+    return <div className={style['console-toolbar']}>
+      <div className={style['console-toolbar-actions']}>
+        <VmConsoleSelector key='console-selector' vmId={match.params.id} consoleId={match.params.console_id} isConsolePage />
```
Suggest to put each attribute on a new line like `VmConsoleSettingsModal`

sjd78 commented on src/components/Toolbar/style.css:
```diff
@@ -5,3 +5,19 @@
 }
+
+.full-width {
+	width: 100%;
```
Nit-pick: tabs instead of spaces?  check indents (same below)

sjd78 commented on src/components/Pages/index.js:
```diff
@@ -150,9 +151,32 @@ const VmCreatePageConnected = connect(
+  (state) => ({
+    vms: state.vms,
+  }),
+  (dispatch) => ({})
```
You can leave the `dispatch` function bit off and only specify the `(state) => {}` function.

bond95 replied:
We are using this form everywhere in code, so I'm leaving it for keeping consistency.

bond95 commented on package.json:
```diff
@@ -68,6 +68,7 @@
   },
   "dependencies": {
     "babel-polyfill": "6.26.0",
+    "@patternfly/react-console": "1.4.6",
```
I tested with latest version and VncConsole work fine, so I leave it in `1.10.22`.

bond95 commented on config/babel.prod.js:
```diff
@@ -5,7 +5,7 @@ module.exports = {
   presets: [
     // Latest stable ECMAScript features
-    require.resolve('babel-preset-latest'),
+    require.resolve('babel-preset-env'),
```
Actually no, this is artifact after Mike.

bond95 commented on src/components/Loader.js:
```diff
@@ -0,0 +1,32 @@
+      wh = 100
+      break
+  }
+  return <div key='infinite-scroll-loader' className={style['loaderBox']}>
```
Removed.

sjd78 replied:
Hum - I'm getting a "key" prop warning from PageRouter now.  Perhaps just make the `key` prop a passthrough prop and put `key='infinite-scroll-loader'` or similar back on where the `Loader` is referenced?

Also, makes sense to me to use this new `Loader` component in both `Vms.js` and in`VmConsole.js`...

sjd78 commented:
@bond95, @gregsheremeta - Try opening a VNC browser console and leave it open for awhile.  Seems to crash my tab every time, I just haven't been able to time it out consistently.

sjd78 commented:
@lwrigh, @gregsheremeta, @bond95 - Thoughts on the breadcrumb title?  Right now it is the VM's console ID.  Maybe just the word "Console" instead?

![screenshot-localhost-3000-2019 02 21-22-49-22](https://user-images.githubusercontent.com/3985964/53218656-f8ab1a00-362a-11e9-8810-4ca939b7368e.png)

sjd78 commented:
@gregsheremeta - Thought ... this could probably be adapted to webadmin via ui-extensions without all that much extra work.

lwrigh commented:
> @lwrigh, @gregsheremeta, @bond95 - Thoughts on the breadcrumb title? Right now it is the VM's console ID. Maybe just the word "Console" instead?
> 
> ![screenshot-localhost-3000-2019 02 21-22-49-22](https://user-images.githubusercontent.com/3985964/53218656-f8ab1a00-362a-11e9-8810-4ca939b7368e.png)

Just having 'Console' as the third breadcrumb makes sense to me.

sjd78 commented on src/components/VmConsole/VmConsoleSelector.js:
```diff
@@ -0,0 +1,97 @@
+        />
+      )
+    }
+    return <div className={style['console-dropdown-box']}>
```
Only suggestion for this component is to define the constants close to where they are used and use whitespace to show a "block level" distinction between the `consoleItems` pushes.

gregsheremeta commented on src/routes.js:
```diff
@@ -48,6 +48,17 @@ export default function getRoutes (vms) {
+        routes: [
+          {
+            path: '/vm/:id/console/:console_id',
+            title: (match) => match.params.console_id,
```
+1, should be 'Console'

sjd78 commented:
@bond95 - Need to be rebased after #967 merged.

sjd78 commented on src/react-console/index.js:
```diff
@@ -0,0 +1,237 @@
+    this.rfb = undefined
+  }
+
+  disconnect = () => {
```
I know over in your patternfly PR (https://github.com/patternfly/patternfly-react/pull/1402#discussion_r258028551) @mareklibra asked why the function change from a class function to an arrow function.

AFAICT, in our project (web-ui), we don't have the same babel level or transforms that auto bind the class functions, but their project does.  So, it is necessary to keep it that way here, but not there.  From a spot check, we're on the same `babel` version, but they are on the `env@^1.6.1` preset and we're on the `latest@6.24.1` preset.

Net result - probably do not need to keep that change on the patternfly-react PR.

bond95 replied:
Yes, I know and problem not only in `this`, but also if look at original code, there are some methods defined like arrow, some - not. And on my opinion this is not consistent.

sjd78 replied:
It was just a thought.  Either way, it'll get resolved in patternfly-react and our local copy of the component can get removed.

sjd78 commented:
@bond95, @gregsheremeta, @lwrigh  - FYI, an easy way to tell if you got all of the strings into the translation bundle is to run the app with the dummy locale.  A script will generate a dummy locale for you that surrounds all of the english with some arrows.  No arrows on a piece of text means the text is data or needs to be localized.

````
  yarn intl:dummy
````

Then open the app with the locale on the URL:
````
   http://localhost:3000?locale=aa
````

For example:
![screenshot-localhost-3000-2019 02 28-13-10-54](https://user-images.githubusercontent.com/3985964/53588175-57582280-3b5a-11e9-93c9-b4187d1cd282.png)

bond95 commented on src/react-console/VncActions.js:
```diff
@@ -0,0 +1,57 @@
+  if (!toolbarContainer) {
+    return toolbar
+  }
+  return document.getElementById(toolbarContainer) &&
```
Actually I don't see big difference, but I prefer more this way, because in this case props more clear and easier to read.

sjd78 replied:
I haven't seen any other patternfly-react or react-bootstrap code do containers/portals this way.  The component always accepts another element to use as the target.

bond95 replied:
According to what I saw, patternfly-react has both id and element, so user can choose which one to use, and it seems to me better solution.

sjd78 replied:
I looked again and found exactly 1 instance of `getElementById` use in a component in patternfly-react.  That use actually looks for a component in one param (scrollable) and if it isn't there, uses a different param (scrollableElementId).  Either way, it's up to Marek in the patternfly-react patch to approve what will be integrated with `react-console` and ultimately adopted here.

gregsheremeta replied:
feel free to get his opinion now ...

sjd78 commented:
@bond95 - I'm still getting #966 when I try to connect to a console of a VM that is powering up.

sjd78 commented on src/components/Loader.js:
```diff
@@ -0,0 +1,32 @@
+      wh = 100
+      break
+  }
+  return <div className={style['loaderBox']}>
```
I just tested again with the latest pushes and I'm not getting the error anymore.  I'll assume it was addresses in a different file.

sjd78 commented on src/index-nomodules.css:
```diff
@@ -211,3 +211,28 @@ a.disabled {
+  height: 100%;
+}
+
+.vnc-console > .toolbar-pf-results {
```
Are these rules needed?

sjd78 commented on src/intl/messages.js:
```diff
@@ -91,6 +91,9 @@ export const messages: { [messageId: string]: MessageType } = {
   details: 'Details',
+  disconnect: { message: 'Disconnect', description: 'Text for disconnect button in noVNC console' },
+  disconectedConsole: 'Disconnected from Console',
+  disconectedConsoleInfo: 'The console has been diconnected. Press the \'Connect\' button to reconnect the console.',
```
The console has been di**s**connected. Press the \'Connect\' button to reconnect the console.

bond95 commented:
> VmConsole's use of VncConsole needs localized text

@sjd78 It is already localized.

sjd78 commented on src/components/VmConsole/index.js:
```diff
@@ -0,0 +1,128 @@
+            port={websocket.get('port')}
+            portalToolbarTo='vm-console-toolbar-sendkeys'
+            onDisconnected={onDisconnected}
+          />
```
- `textConnecting` -- the `loaderText` needs localization
  - `textDisconnected` :heavy_check_mark: -- ah ha... it isn't needed since rendering the disconnected state is handled in `VmConsole`
  - `textSendShortcut` :heavy_check_mark: 
  - `textCtrlAltDel` :heavy_check_mark: 
  - `textDisconnect` :heavy_check_mark:

lwrigh commented:
> @bond95, @gregsheremeta, @lwrigh - FYI, an easy way to tell if you got all of the strings into the translation bundle is to run the app with the dummy locale. A script will generate a dummy locale for you that surrounds all of the english with some arrows. No arrows on a piece of text means the text is data or needs to be localized.
> 
> ```
>   yarn intl:dummy
> ```
> Then open the app with the locale on the URL:
> 
> ```
>    http://localhost:3000?locale=aa
> ```
> For example:
> ![screenshot-localhost-3000-2019 02 28-13-10-54](https://user-images.githubusercontent.com/3985964/53588175-57582280-3b5a-11e9-93c9-b4187d1cd282.png)

Ohh this is a cool trick!

sjd78 commented:
ci please build

leistnerova commented:
I have few issues:
1. Console select box is not closed after clicking on browser option. Go to browser console, select other console option - download page is displayed - select back browser console.
![screenshot from 2019-03-07 11-20-40](https://user-images.githubusercontent.com/29575480/53950311-292f9100-40cc-11e9-983f-c3b5fa0b2c7a.png)

2. icon of disconnect has bad quality
![screenshot from 2019-03-07 11-26-12](https://user-images.githubusercontent.com/29575480/53950341-39477080-40cc-11e9-9cdf-261eb9ab7e45.png)

3. link to Remote Desktop Connection Settings is disabled for VNC console, for VNC console (browser) is enabled, maybe it is switched or it should be enabled for both. And if I'm not mistaken spice can be also opened with virt-viewer.
![screenshot from 2019-03-07 11-29-52](https://user-images.githubusercontent.com/29575480/53950433-6c89ff80-40cc-11e9-8d12-128125334abb.png)

4. rfe - User opened browser console, when VM is shutdown console changes to disconnect. When user attempts to open console again (no matter what type - spice/vnc/browser) and VM is in Down state, close the console browser and go back to VM detail. There are only errors with retrieving console details now.

There was also suspicious behavior with suspended VM and vnc doesn't work for me in application. I need to investigate that more and I'll write other comment then.

Tested on FF 65.0.1 with ovirt-engine-4.3.2.1-0.0.master.20190305140204.git3649df7.el7.noarch

leistnerova commented:
Suspended VM works as expected, it was my FF overload not portal fault. VNC in application doesn't work also from admin portal, so not VM portal related.

bond95 commented:
@leistnerova 

> Console select box is not closed after clicking on browser option. Go to browser console, select other console option - download page is displayed - select back browser console.

Fixed

> icon of disconnect has bad quality

This is strange because even when I zoom it, it keep be smooth
![image](https://user-images.githubusercontent.com/3332176/54206755-3b00a200-44d9-11e9-9db8-f0b6ac95204c.png)

>link to Remote Desktop Connection Settings is disabled for VNC console, for VNC console (browser) is enabled, maybe it is switched or it should be enabled for both. And if I'm not mistaken spice can be also opened with virt-viewer.

I'd like to hear what other think @gregsheremeta @sjd78 @lwrigh 

> rfe - User opened browser console, when VM is shutdown console changes to disconnect. When user attempts to open console again (no matter what type - spice/vnc/browser) and VM is in Down state, close the console browser and go back to VM detail. There are only errors with retrieving console details now.

Now when VM shutdown it redirect user to VM dashboard page.

bond95 commented:
@gregsheremeta @sjd78 @lwrigh 
Updated images:
![image](https://user-images.githubusercontent.com/3332176/54432004-37168f00-4728-11e9-9536-41f03d262065.png)

Renamed Console Settings Modal -> Console Instructions and enabled to VNC and SPICE consoles:
![image](https://user-images.githubusercontent.com/3332176/54432046-5d3c2f00-4728-11e9-9ece-dabea6c407ca.png)

Removed Download button:
![image](https://user-images.githubusercontent.com/3332176/54432091-8492fc00-4728-11e9-8930-b69107f0c4fa.png)

lwrigh commented:
> @gregsheremeta @sjd78 @lwrigh
> Updated images:
> ![image](https://user-images.githubusercontent.com/3332176/54432004-37168f00-4728-11e9-9536-41f03d262065.png)
> 
> Renamed Console Settings Modal -> Console Instructions and enabled to VNC and SPICE consoles:
> ![image](https://user-images.githubusercontent.com/3332176/54432046-5d3c2f00-4728-11e9-9ece-dabea6c407ca.png)
> 
> Removed Download button:
> ![image](https://user-images.githubusercontent.com/3332176/54432091-8492fc00-4728-11e9-8930-b69107f0c4fa.png)

These updates are looking great! Thank you for updating it.

gregsheremeta commented on src/intl/messages.js:
```diff
@@ -394,8 +394,8 @@ export const messages: { [messageId: string]: MessageType } = {
   },
-  remoteDesktopConnectionSettings: 'Remote Desktop Connection Settings',
-  remoteDesktopViewerConnection: 'Remote Desctop Viewer Connection',
+  remoteDesktopConnectionInstructions: 'Console Instructions',
```
rename `vmConsoleConnectionInstructions`

gregsheremeta commented on src/intl/messages.js:
```diff
@@ -394,8 +394,8 @@ export const messages: { [messageId: string]: MessageType } = {
-  remoteDesktopConnectionSettings: 'Remote Desktop Connection Settings',
-  remoteDesktopViewerConnection: 'Remote Desctop Viewer Connection',
+  remoteDesktopConnectionInstructions: 'Console Instructions',
+  remoteDesktopViewerConnection: 'Remote Desktop Viewer Connection',
```
rename `consoleInstructions`

string should match the title, which is now `Console Instructions`

gregsheremeta commented on src/intl/messages.js:
```diff
@@ -394,8 +394,8 @@ export const messages: { [messageId: string]: MessageType } = {
-  remoteDesktopViewerConnection: 'Remote Desctop Viewer Connection',
+  remoteDesktopConnectionInstructions: 'Console Instructions',
+  remoteDesktopViewerConnection: 'Remote Desktop Viewer Connection',
   remoteDesktopViewerAvailable: 'Remote Desktop Viewer is available for most operating systems. To install it, search for it in GNOME Software or run the following:',
```
rename `remoteViewerAvailable`

The thick client's name is `Remote Viewer` (no `Desktop`) so remove `Desktop`
(The term `Remote Desktop` is easy to confuse with MS RDP, so I want to be really careful)

gregsheremeta commented:
ci build please

gregsheremeta commented:
all yours, @leistnerova  :)

leistnerova commented:
Main functionality works fine, I have also some little issues. Lets merge it and improve later.

leistnerova commented:
Ah, I forgot about pools again, sorry... so fix Console button in pool and then merge.
![Screenshot from 2019-03-21 10-22-58](https://user-images.githubusercontent.com/29575480/54743053-62cfc400-4bc3-11e9-8552-12957558f004.png)

leistnerova commented:
After talking to Greg there are other issue that should be also fixed:

1. When VM in reboot state, console should stay normally working. 
Steps to reproduce:
  1. run VM with VNC console enabled
  2. go to VNC console (browser)
  3. click on Send Key -> Ctrl+Alt+Del
  4. stay in the console
Console starts to reboot, you see boot menu and after automate refresh the view goes to VM detail and user is not able to go back to console again.

2. After opening VNC browser console, the console window should get focus. Now user have to click in the window to be able to start typing.

Sorry for the chaos, my mistake I understood wrong the purpose of merging it now.

gregsheremeta commented:
I also noticed the reboot issue @leistnerova is describing.

Things I'd like to test:
1. windows 7 vm
2. windows 8 vm
3. windows 10 vm
4. macOS client, just to make sure novnc works in safari

bond95 commented:
@leistnerova fixed
![image](https://user-images.githubusercontent.com/3332176/54764581-5e71ce00-4bf8-11e9-91fc-5ffcfb435102.png)

sjd78 commented:
> I also noticed the reboot issue @leistnerova is describing.
> 
> Things I'd like to test:
>
> 4. macOS client, just to make sure novnc works in safari

Just tried it in Safari on OSX and the console looks like it connects, and there are no errors on the javascript console, but no display.
<img width="1440" alt="safari-vnc-console" src="https://user-images.githubusercontent.com/3985964/54799077-f4761a80-4c31-11e9-98ad-d877654b1f48.png">

bond95 commented:
> Just tried it in Safari on OSX and the console looks like it connects, and there are no errors on the javascript console, but no display.

@sjd78 fixed, please, check

![image](https://user-images.githubusercontent.com/3332176/54933337-c12ad880-4f1c-11e9-814a-43ad6f80c523.png)

sjd78 commented on src/components/VmConsole/index.js:
```diff
@@ -110,7 +112,9 @@ class VmConsole extends React.Component {
             toolbarContainer='vm-console-toolbar-sendkeys'
+            containerId={NOVNC_CONTAINER_ID}
             onDisconnected={onDisconnected}
+            onConnected={() => $(`#${NOVNC_CONTAINER_ID} canvas`).focus()}
```
Does `document.querySelector()` not work for that selector?

sjd78 commented:
> > Just tried it in Safari on OSX and the console looks like it connects, and there are no errors on the javascript console, but no display.
> 
> @sjd78 fixed, please, check

Looks good on my end:
<img width="1332" alt="working novnc" src="https://user-images.githubusercontent.com/3985964/54937045-130e3700-4efa-11e9-96d7-05bd2b257fab.png">

leistnerova commented:
I've found other issues
1. When admin opens console on VM with no guest tools installed. Non admin user in VM portal doesn’t get any error on opening console. Only warning that it is already in use.
Web console contains:

debug Ajax failed: {"readyState":4,"responseText":"{\n \"fault\" : {\n \"detail\" : \"[Guest agent non responsive]\",\n \"reason\" : \"Operation Failed\"\n },\n \"status\" : \"failed\"\n}","responseJSON":{"fault":{"detail":"[Guest agent non responsive]","reason":"Operation Failed"},"status":"failed"},"status":400,"statusText":"Bad Request"} transport.js:104:6

2. Missing certificate installation information from https://www.ovirt.org/documentation/admin-guide/virt/console-client-resources.html

3. Sending Ctrl+Alt+Del removes focus from the console. And so then pressing ESC (e.g. for boot menu) exits the console to VM detail

Tested in FF 65.0.1 and Google Chrome 73.0
with ovirt-engine-4.4.0-0.0.master.20190318180517.git576124b.el7.noarch

Edge seems also to work ok. 
I have only problem sometimes that mouse in console is slower than my normal mouse. But I assume that is problem of my vnc driver and installed windows not VM portal.

bond95 commented:
After bad handshake:
![image](https://user-images.githubusercontent.com/3332176/55150253-0b45d100-514c-11e9-8477-efd8380e4cda.png)

Regular disconnect:
![image](https://user-images.githubusercontent.com/3332176/55150335-39c3ac00-514c-11e9-943a-c6dc94ae9a6f.png)

@lwrigh @gregsheremeta @sjd78 @leistnerova WDYT?

lwrigh commented:
> After bad handshake:
> ![image](https://user-images.githubusercontent.com/3332176/55150253-0b45d100-514c-11e9-8477-efd8380e4cda.png)
> 
> Regular disconnect:
> ![image](https://user-images.githubusercontent.com/3332176/55150335-39c3ac00-514c-11e9-943a-c6dc94ae9a6f.png)
> 
> @lwrigh @gregsheremeta @sjd78 @leistnerova WDYT?

I would recommend changing the text for the bad handshake error to, 'The console has been disconnected because of a security error. For more information please refer to the [console manual](url). Press the 'Connect' button to reconnect the console.'

bond95 commented:
@lwrigh Done
![image](https://user-images.githubusercontent.com/3332176/55158392-12c2a580-515f-11e9-8626-d9473279f71b.png)

leistnerova commented:
Issue in Microsoft Edge 42 - Clicking on VNC or SPICE console (not browser) downloads the file but doesn't close the console dropdown menu. In main grid and also VM detail.

bond95 commented:
@leistnerova fixed, please, test.

leistnerova commented:
ci build please

leistnerova commented:
I've found another issue:
VMs drop-down menu in VM grid view shows only Console and nothing happens when user clicks on it. After going to detail and back, it shows correctly SPICE,VNC consoles and it works fine.

![Screenshot from 2019-04-09 11-26-56](https://user-images.githubusercontent.com/29575480/55789445-a6c73200-5aba-11e9-9d0e-003b785e1325.png)

Steps to reproduce:
1. log into VM portal with non-admin user
2. add permissions on new running VM to the user in admin portal
3. reload VM main grid view in VM portal

bond95 commented:
@sjd78 please, review.

sjd78 commented on src/sagas/login.js:
```diff
@@ -234,6 +236,15 @@ function* initialLoad () {
 }
 
+function* fetchWebsocketInformation () {
+  const data = yield callExternalAction('getOption', Api.getOption, getOption('WebSocketProxy', 'general', ''))
```
Do we know if it is possible to connect NoVNC if the websocket proxy isn't available?

If websocket proxy is required, do we have handling on the web console connection if it is not defined?  (I haven't read through it in enough detail to remember at the moment)

bond95 replied:
> Do we know if it is possible to connect NoVNC if the websocket proxy isn't available?

Generally `WebSocketProxy` is a URL through which client will connect to NoVNC console, if it is unavailable, then client can't connect to NoVNC.

> If websocket proxy is required, do we have handling on the web console connection if it is not defined? (I haven't read through it in enough detail to remember at the moment)

Only what I found about it: `If the console does not start up automatically, this empty state will appear. The user can click the primary or secondary ‘Connect’ button to start up the console.`

I think handling of this situation can be moved to separate issue, because this PR is already too big, so it is better to split next issues.

sjd78 replied:
Detecting a missing or empty websocket proxy url (or a simple sanity test to make sure it looks right) followed by an error message or a toast would probably be enough.  It is related to if the proxy is down, or if the url has an error, or if the engine root cert isn't approved on the browser and wss:// is blocked.  They're all instance of the websocket not being right and that the connection will fail because of that.

@lwrigh should weigh in on what kind of error display/response is reasonable.  Toast, error message box?  Distinct messages for each kind of problem or a generic "can't connect to your VM's console due to proxy errors" message?  I think hiding the browser console button would be confusing.  @sgratch (or even @gregsheremeta) - Thoughts?

@leistnerova made a good point at standup that with little time pressure currently, handling this edge case in this PR is probably worth it (even if it is super annoying).

lwrigh replied:
I would definitely avoid hiding the browser console button, that would make for a really confusing experience. I think in this case it would make the most sense to feature the message dialog component (https://www.patternfly.org/pattern-library/communication/message-dialog/#design) and feature a message about the proxy error. Would like this make sense as an error message,"Cannot connect to [Name of VM]'s console because of a proxy error. Please try again.". ? The actions on the dialog could be 'Cancel' and 'Try Again'.

bond95 commented:
@lwrigh @sjd78 
![image](https://user-images.githubusercontent.com/3332176/56127716-16876200-5f7e-11e9-943c-cd7c505f49c0.png)
![image](https://user-images.githubusercontent.com/3332176/56127745-230bba80-5f7e-11e9-8663-5cf9d0d0b49a.png)

michalskrivanek commented:
How does the empty URL mean certificate is not ok? If it's just looking at the WebSocketProxy option that's insufficient (to say the least;)

bond95 commented:
> How does the empty URL mean certificate is not ok? If it's just looking at the WebSocketProxy option that's insufficient (to say the least;)

@michalskrivanek I don't really understand question. It doesn't tell about certificate, it notify user that proxy is unavailable.

michalskrivanek commented:
> @michalskrivanek I don't really understand question. It doesn't tell about certificate, it notify user that proxy is unavailable, because URL is empty.

and what does that mean? why/when is it unavailable?

bond95 commented:
> > @michalskrivanek I don't really understand question. It doesn't tell about certificate, it notify user that proxy is unavailable, because URL is empty.
> 
> and what does that mean? why/when is it unavailable?

If user doesn't configurated websocket proxy in ovirt-engine.

sjd78 commented:
@bond95 , @leistnerova - How is this PR looking?  Latest round of code changes look ok.  Does it just need another round of testing?

leistnerova commented:
I'll do a quick test of last changes, not the whole testing.

leistnerova commented:
ci build please

bond95 commented:
@sjd78 @lwrigh WDYT?

<img width="1307" alt="Screenshot 2019-05-06 at 11 07 46" src="https://user-images.githubusercontent.com/3332176/57216128-452db100-6fef-11e9-81b4-35fc85763753.png">

lwrigh commented:
I would suggest changing the text to read, "Cannot connect to websocket proxy server. Please check your websocket proxy certificate or ask your administrator for help. For further information please refer to the console manual(linked). Press the 'Connect' button to reconnect to the console.".

bond95 commented:
@lwrigh Done
<img width="1433" alt="Screenshot 2019-05-07 at 11 18 00" src="https://user-images.githubusercontent.com/3332176/57288507-dff2c200-70b9-11e9-9979-03ca9d009633.png">

lwrigh commented:
Nice looks good to me!

leistnerova commented:
LGTM

sjd78 commented:
Just noticed while redoing a demo video: At least on Chrome, if you click directly from the console to the details in the breadcrumbs, the charts on the details page don't render:
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/3985964/57489101-0c0d6f00-7283-11e9-93b4-69ba5d5df029.gif)

My best guess is that c3 gets broken somewhere when you don't explicitly click the disconnect button.  Eventually my browser tab (Chrome) will straight crash as well.

@bond95, @lwrigh, @leistnerova - Can you guys reproduce?
@sgratch, @michalskrivanek - Punt to an issue or resolve in this PR (assuming it can be reproduced)?

bond95 commented:
> @bond95, @lwrigh, @leistnerova - Can you guys reproduce?

@sjd78 Unfortunately, no, everything works fine for me in Chrome Version 73.0.3683.86 (Official Build).
![Peek 2019-05-10 11-23](https://user-images.githubusercontent.com/3332176/57517044-3fe5a480-7316-11e9-985d-a3f809b5ffdf.gif)

sjd78 commented:
> @sgratch, @michalskrivanek - Punt to an issue or resolve in this PR (assuming it can be reproduced)?

I'll do some additional testing permutations (Chrome on a Mac connecting to dev server web-ui and connecting to web-ui installed to my dev ovirt-engine instance).  If it still comes up, I'll log a separate issue.  I'm ok with merging this as-is.

bond95 commented:
> I'll do some additional testing permutations (Chrome on a Mac connecting to dev server web-ui and connecting to web-ui installed to my dev ovirt-engine instance). If it still comes up, I'll log a separate issue. I'm ok with merging this as-is.

Great, thank you.

leistnerova commented:
@sjd78 I also did not reproduce is, tested in Chrome 74.0 on Fedora

sjd78 commented:
> > @sgratch, @michalskrivanek - Punt to an issue or resolve in this PR (assuming it can be reproduced)?
> 
> I'll do some additional testing permutations (Chrome on a Mac connecting to dev server web-ui and connecting to web-ui installed to my dev ovirt-engine instance). If it still comes up, I'll log a separate issue. I'm ok with merging this as-is.

I tested Chrome 74.0 on Mac against the web-ui dev server and against web-ui installed to my dev ovirt-engine and was able to reproduce the c3 no charts problem and the browser tab crashing problem.  The tab appears to crash because of a massive memory leak.  I also tested with Firefox on fedora against my dev ovirt-engine with web-ui installed and recreated the memory leak there.  Firefox's tab doesn't crash as politely as the Chrome tab.

Here is a video of the Chrome 74.0 on Mac C3 fail followed my memory leak and tab crash: 
https://youtu.be/ZrOGGyPQpPA

leistnerova commented:
It seems I have the reproduction steps:
1. have running VM
2. check browser console, go back to VM detail
3. Shutdown with Force (normal shutdown seems to work)
4. Run VM again
5. go to browser console, wait a while
6. when going back to VM detail, charts already loaded
7. go to browser console, back to detail
8. charts broken

It doesn't happen 100%, but mostly. Reproduced in Google Chrome 74.0 and also Firefox 66.0

michalskrivanek commented:
@leistnerova thanks! @sjd78 does the console code terminate properly when VM is shut down? if not...could be novnc bug or our code around it. Generally as soon as VM is down the console should die with it.

bond95 commented:
> @leistnerova thanks! @sjd78 does the console code terminate properly when VM is shut down? if not...could be novnc bug or our code around it. Generally as soon as VM is down the console should die with it.

Console code must be terminated with VmConsole component. When user go to detail page from console page, VmConsole is unmounting and console terminate as well. According to my experiments problem can be in charts, because this bug cause frequent update of charts, which can be cause of growing memory usage.

sgratch commented:
@bond95 @sjd78 if both #966 and #490 is handled by this PR then can we please merge this as discussed and open a new issue for https://github.com/oVirt/ovirt-web-ui/pull/956#issuecomment-496546223 ?

bond95 commented:
@sgratch yes, of course, I'll merge this.
